PSA: Phase 1: Test using opaque private ECDSA key in TLS client

mpg opened the pull request:
## Description

### Task description

Test usage of PK_OPAQUE_PSA in ssl_set_own_cert()

* adapt `ssl_client/server2.c`
* might need internal adaptations in TLS to work
* add test using that option to `ssl-opt.sh`
* make sure it's run in `all.sh`

Scope restriction: the ECDSA part of the ECDHE-ECDSA key exchange, TLS 1.2

### Execution of the task

* did only `ssl_client2.c` as doing server too would require some code duplication while waiting for #1272, and we care mostly about client side so far.
* turned out no internal adaptation in the TLS module was needed
* test added to `ssl-opt.sh` as planned
* making sure the added test (in fact, all available tests) is run in `all.sh` in a PSA-enabled build is part of the base PR so nothing more's needed here

**Note:** This branch is based on #2162 - only new commits compare to that PR need reviewing here.

## Status
**READY** - though it will need rebasing before merge.

## Requires Backporting
NO

## Migrations
NO

## Steps to test or reproduce

* build with `MBEDTLS_USE_PSA_CRYPTO` enabled in`config.h` (as well as its dependencies)
* run `tests/ssl-opt.sh -f Opaque`

mpg commented:
cc: @gilles-peskine-arm - I'm not requesting a formal review from you, but if you feel like checking how we're using your APIs, feel free to have a look at a92c360 and let us know if you see any problem.

mpg commented:
@hanno-arm Thanks for your review! I believe I've addressed your feedback, please review again!

mpg commented:
@hanno-arm @AndrzejKurek I've rebased over the lastest #2162 (rebased on #2184) so that I could eliminated redundancies with #2184. I've also simplified the history while at it. The previous history is still available at: https://github.com/mpg/mbedtls/tree/iotssl-2574-pk-opaque-tls-0 Please review again.

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -726,6 +737,8 @@ int main( int argc, char *argv[] )
             opt.crt_file = q;
         else if( strcmp( p, "key_file" ) == 0 )
             opt.key_file = q;
+        else if( strcmp( p, "key_opaque" ) == 0 )
```
I don't think we should parse an option if it's not advertised in the `USAGE` string. Could you guard this by `MBEDTLS_USE_PSA_CRYPTO`, too?

mpg replied:
We do parse all the other options even if they're not advertised in the usage string, and initially I was inclined to just to the same for homogeneity's sake. However, I agree that refusing to parse options we don't support would be a good think, so we can start adding that to new options and fix the old ones later on.

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -529,6 +533,67 @@ int idle( mbedtls_net_context *fd,
+    mbedtls_pk_free( pk );
+    mbedtls_pk_init( pk );
+
+    return( mbedtls_pk_setup_opaque( pk, key ) );
```
Minor: There's an inconsistency in error-handling here: `mbedtls_pk_setup_opaque()` returns a meaningful error code, while earlier branches return with `-1`. I'd prefer to use `MBEDTLS_ERR_PK_HW_ACCEL_FAILED` earlier.

mpg replied:
Good idea. Doing that.

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -1325,11 +1393,16 @@ int main( int argc, char *argv[] )
     if( opt.key_opaque != 0 )
     {
-        /* coming soon: load key to key slot */
+        if( ( ret = wrap_key_as_opaque( &pkey, &key_slot ) ) != 0 )
```
NB: That's a neat way to adapt the code with just a single line of change! :+1:

mpg replied:
Thanks :)

mpg commented:
@hanno-arm Thanks for your review! I believe I've addressed your feedback. Please review again.

Note: I did *not* rebase over the most recent #2162 because the recent changes in that branch do not affect anything here, so it wouldn't bring any value now, and everything will need rebasing later anyway.

AndrzejKurek commented on programs/ssl/ssl_client2.c:
```diff
@@ -134,9 +135,16 @@ int main( void )
 #endif /* MBEDTLS_X509_CRT_PARSE_C */
+#if defined(MBEDTLS_USE_PSA_CRYPTO) && defined(MBEDTLS_X509_CRT_PARSE_C)
+#define USAGE_OPAQUE \
+    "    key_opaque=%%d       Handle your private key as if it were opaque\n" \
```
I'd rather say "As if it was". Was is used if the situation is probable, and "were" when it's pretty much impossible (simplyfying). 
If I were a rich man... ;)

gilles-peskine-arm replied:
That's not a rule I've ever heard of and I suspect that you're confusing it with a slightly different rule, which is that it's *were* (subjunctive) is for hypotheticals and counter-factuals, whereas *was* is for something that plausibly happened in the past. By this rule, the correct word is *were*, since the private key is not in fact opaque.

However the subjunctive in English is somewhat old-fashioned. It's more common in the US, but losing ground. I'd use *was*, but just because it's more likely to be understood, not because of some difference in meaning.

mpg replied:
Well, currently it is impossible to provide a truly opaque key... well, I think it depends on your definition of possible, because of course we could add such an option later, but that's currently not the case. I think it's similar to "if I were rich": it's possible that this might become true in the future, but we're insisting that it's currently not the case at all. I think both could make sense for this option, but I felt more like insisting that the private key isn't really opaque so far.

AndrzejKurek commented on programs/ssl/ssl_client2.c:
```diff
@@ -337,6 +346,7 @@ struct options
     const char *ca_path;        /* the path with the CA certificate(s) reside */
     const char *crt_file;       /* the file with the client certificate     */
     const char *key_file;       /* the file with the client key             */
+    int key_opaque;             /* handle private key as if it were opaque  */
```
Same as above

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -515,6 +533,67 @@ int idle( mbedtls_net_context *fd,
+ *
+ * Currently only works for EC private keys.
+ */
+static int wrap_key_as_opaque( mbedtls_pk_context *pk, psa_key_slot_t *slot )
```
@mpg @AndrzejKurek: As this function is also used in #2197, could we move it to the parent PR #2162 for both this and #2197 as a helper function in `pk_internal.h`?

mpg commented:
@hanno-arm @AndrzejKurek I rebased on #2162 (itself rebased on recent #2206), used the utility function that I recently moved from here to #2162, and cleaned up the history. Please review again.

mpg commented:
A local run of `all.sh -k -r` passed on cc93b01 except for the following independent issues:

- Mbed Crypto exported: will be removed soon
- test that triggers a GnuTLS bug: addressed by #2209

AndrzejKurek commented on programs/ssl/ssl_client2.c:
```diff
@@ -1327,11 +1330,17 @@ int main( int argc, char *argv[] )
     {
-        /* coming soon: load key to key slot */
+        if( ( ret = mbedtls_pk_wrap_as_opaque( &pkey, &key_slot,
+                                               PSA_ALG_SHA_256 ) ) != 0 )
```
I would rather suggest defining it closer to the key or even get it from parsing and using the translating utility rather than hardcoding.

mpg replied:
In the long run I agree, but for the time being I'm going to consider it out of scope of this PR.

The main purpose was to make sure that opaque keys can be used for client authentication, ie that there wasn't any need to change code in the TLS module in order to adapt to opaque keys (same for your CSR generation PR, which made it clear that some change was needed in pkwrite, except here it turned out no change was needed). That purpose can be achieved with a hardcoded hash.

But yeah, I assume a subsequent PR is going to change that.

mpg commented on library/pk.c:
```diff
@@ -139,6 +143,38 @@ int mbedtls_pk_setup( mbedtls_pk_context *ctx, const mbedtls_pk_info_t *info )
+
+    /* Current implementation of can_do() relies on this. */
+    if( ! PSA_KEY_TYPE_IS_ECC_KEYPAIR( type ) )
+        return( MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE) ;
```
Because "in principle" (and hopefully in practice in the near future) the function should support RSA too... I agree `BAD_INPUT_DATA` would make sense too for now, but I'd rather work on removing that restriction than worry too much about picking the best return code.

mpg commented on library/pk.c:
```diff
@@ -501,4 +539,65 @@ mbedtls_pk_type_t mbedtls_pk_get_type( const mbedtls_pk_context *ctx )
+    /* set policy */
+    psa_key_policy_init( &policy );
+    psa_key_policy_set_usage( &policy, PSA_KEY_USAGE_SIGN,
+                                       PSA_ALG_ECDSA(hash_alg) );
```
Thanks for the hint! I'll do that.

mpg replied:
Hmm, wait. What happens if we set the policy to `PSA_ALG_ECDSA_DETERMINISTIC` here and someone later tries to make a signature with `PSA_ALG_ECDSA`? Or the contrary?

gilles-peskine-arm replied:
The algorithms have to match. But to emulate the transparent behavior, you'd pick the same algorithm again when you sign anyway.

A bigger issue is that for deterministic ECDSA you need to pick the hash in advance, whereas for randomized ECDSA you can cheat. We've talked about the hash commitment problem before and I don't have a solution yet.

mpg replied:
I think we really should have a list of allowed algorithms, rather than just one. I mean we can already allow multiple usages, why would multiple algorithms for the same usage be a bigger issue? For example, it's not clear to me that using the same RSA key with PKCS#1 v1.5 and v2 would be any worse than using it both for signing and decrypting.

mpg commented on include/mbedtls/pk.h:
```diff
@@ -234,6 +244,38 @@ void mbedtls_pk_restart_free( mbedtls_pk_restart_ctx *ctx );
+ * \note            This function replaces mbedtls_pk_setup() for contexts
+ *                  that wrap a (possibly opaque) PSA key slot instead of
+ *                  storing and manipulating the key material directly.
+ *
```
The documentation of `MBEDTLS_USA_PSA_CRYPTO` has such a warning. And since it's off by default, the theory is that you'll read the warning before you enable it.

mpg commented on library/pk_wrap.c:
```diff
@@ -716,4 +724,182 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+        --len;
+    }
+
+    /* this is only reached if the signature was invalid */
```
I'm sorry, I don't get the nuance between incorrect and malformed here. Note that per the standard, neither r nor s are allowed to be 0.

gilles-peskine-arm replied:
To me a valid signature of m is a signature rs such that the verify function V(m,rs) returns true. If V(m1,rs1) and V(m2,rs2) are true and m1 ≠ m2 then rs1 is a valid signature of m1 but an invalid signature when you're checking m2. But “this is only reached …” doesn't apply to rs2. While it is logically true that it applies only if the signature is invalid, it doesn't apply to every invalid signature, only to certain very specific inputs that technically aren't signatures at all.

mpg commented on library/pk_wrap.c:
```diff
@@ -716,4 +724,182 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+     * that information. Assume that the buffer is large enough for a
+     * maximal-length signature with that key (otherwise the application is
+     * buggy anyway). */
+    status = psa_get_key_information( *key, NULL, &bits );
```
The key type is assumed to be ECC because otherwise `mbedtls_pk_setup_opaque()` wouldn't have allowed it, and we're assuming users don't create opaque contexts any other way than using this function. Oh, wait, I should enforce that in `mbedtls_pk_setup()`, because that looks like an easy mistake to do.

hanno-arm commented:
@mpg I took the liberty to rebase this on top of the current `psa-integration-utilities` and changed the target branch to `development-psa`.
